Add new sample for shader debugprintf

SaschaWillems opened the pull request:
## Description

This PR adds a new sample showing how to use debugprintf in shaders. It also includes a tutorial.

Note: This also does some changes to the framework required for samples that do custom instance creation.

Tested on Windows 11 with an NVIDIA RTX 4070

Fixes #926
Fixes #985

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#Copyright-Notice-and-License-Template)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [x] My changes do not add any regressions
- [x] I have tested every sample to ensure everything runs correctly
- [x] This PR describes the scope and expected impact of the changes I am making

 Note: The Samples CI runs a number of checks including:
 - [x] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
 - [x] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#General-Requirements)

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [x] I have tested the sample on at least one compliant Vulkan implementation
- [x] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/CONTRIBUTING.adoc#General-Requirements)
- [x] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [x] Any dependent assets have been merged and published in downstream modules
- [x] For new samples, I have added a paragraph with a summary to the appropriate chapter in the [samples readme](https://github.com/KhronosGroup/Vulkan-Samples/blob/main/samples/README.adoc)
- [x] For new samples, I have added a tutorial README.md file to guide users through what they need to know to implement code using this feature. For example, see [conditional_rendering](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/samples/extensions/conditional_rendering)
- [x] For new samples, I have added a link to the [Antora navigation](https://github.com/KhronosGroup/Vulkan-Samples/blob/main/antora/modules/ROOT/nav.adoc) so that the sample will be listed at the Vulkan documentation site

SaschaWillems committed:
Added new shader debzgprrintf sample
Work-in-progress

SaschaWillems committed:
Note on descriptor size
Minor cleanup

SaschaWillems committed:
Display shaderdebutprintf output in UI

SaschaWillems committed:
Proper extension structure chaining

SaschaWillems committed:
Restructure code
Use VK 1.2 to avoid a bug with recent Vulkan SDKs

SaschaWillems committed:
Minor adjustments

SaschaWillems committed:
Use VK 1.1, clean up messages for UI display

gary-sweet reviewed (changes requested):
We don't support debugPrintfEXT but I'm still seeing a couple of issues when I run this PR.

I get a validation error at the start:
`VUID-vkCreateInstance-ppEnabledExtensionNames-01388(ERROR / SPEC): msgNum: -437968512 - Validation Error: [ VUID-vkCreateInstance-ppEnabledExtensionNames-01388 ] | MessageID = 0xe5e52180 | vkCreateInstance(): pCreateInfo->ppEnabledExtensionNames[0] Missing extension required by the instance extension VK_KHR_display: VK_KHR_surface. The Vulkan spec states: All required extensions for each extension in the VkInstanceCreateInfo::ppEnabledExtensionNames list must also be present in that list (https://www.khronos.org/registry/vulkan/specs/1.3-extensions/html/vkspec.html#VUID-vkCreateInstance-ppEnabledExtensionNames-01388)`

It terminates with an error about `VK_KHR_get_physical_device_properties2`:
```
[error] [framework/platform/platform.cpp:169] Error Message: Couldn't request feature from device as VK_KHR_get_physical_device_properties2 isn't enabled!
[error] [framework/platform/platform.cpp:170] Failed when running application shader_debugprintf
Segmentation fault
```
and then segfaults in `vkDestroySurfaceKHR()` called from `VulkanSample::~VulkanSample()`

SaschaWillems commented:
Thanks for noticing. The framework seems to add `VK_KHR_surface` explicitly and via glfw (which looks wrong on e.g. Windows as it shows up twice). I guess on your platform the implicit addition via glfw doesn't happen, hence it's missing. Will fix that.

SaschaWillems committed:
Enable instance extensions

SaschaWillems committed:
Merge branch 'main' into shader_printf

SaschaWillems committed:
Started working on documentation/tutorial

SaschaWillems committed:
Fixed link

SaschaWillems committed:
Merge branch 'main' into shader_printf

SaschaWillems committed:
Minor reordering
Easier to document

SaschaWillems committed:
Tutorial fpr shader printf

SaschaWillems commented:
I have added a tutorial for this sample. It's now ready for review.

asuessenbach commented on samples/extensions/shader_debugprintf/README.adoc:
```diff
@@ -0,0 +1,197 @@
+
+This sample demonstrates the usage of the https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_KHR_shader_non_semantic_info.html[VK_KHR_shader_non_semantic_info] for adding means to output debug messages from SPIR-V shaders using a print function at runtime.
+
+It allows you to output invocation related debug messages from a give shader stage and can help to find problems or bugs together with a graphics debugger like a https://www.renderdoc.org[RenderDoc]. But it's also possible to display that data in your application (e.g. a debug window or the command line), although this has some limitations like a fixed buffer size that all messages need to fit in ...
```
typo: give -> given
typo: like a -> like

asuessenbach commented on samples/extensions/shader_debugprintf/README.adoc:
```diff
@@ -0,0 +1,197 @@
+
+== Setup
+
+Shader printf is a feature of the validation layers. So the firs step is to enable the `VK_LAYER_KHRONOS_validation` layer at instance creation. In addition to the layer(s) to be enabled we also need to provide `VkValidationFeaturesEXT` with the `VK_VALIDATION_FEATURE_ENABLE_DEBUG_PRINTF_EXT` feature enabled:
```
typo: firs -> first

asuessenbach commented on samples/extensions/shader_debugprintf/README.adoc:
```diff
@@ -0,0 +1,197 @@
+    std::vector<VkValidationFeatureEnableEXT>  validation_feature_enables = {VK_VALIDATION_FEATURE_ENABLE_DEBUG_PRINTF_EXT};
+
+    VkValidationFeaturesEXT validation_features{VK_STRUCTURE_TYPE_VALIDATION_FEATURES_EXT};
+    validation_features.enabledValidationFeatureCount = 1;
```
instead of plain `1`, we should use `static_cast<uint32_t>(validation_feature_enables.size())`
Or with just one feature enabled, maybe just use a plain `VkValidationFeatureEnableEXT enable_printf = VK_VALIDATION_FEATURE_ENABLE_DEBUG_PRINTF_EXT;`

SaschaWillems replied:
I'll use `size()`. For people learning from our samples it imo makes it easier if they see that this can be a vector of enables.

asuessenbach commented on samples/extensions/shader_debugprintf/README.adoc:
```diff
@@ -0,0 +1,197 @@
+
+[NOTE]
+====
+Shader printf messages are of `info` type, so in order for them to be passed to the callback you need to enable set the `VK_DEBUG_UTILS_MESSAGE_SEVERITY_INFO_BIT_EXT` `messageSeverity` flag.
```
typo: to enable set -> to enable

asuessenbach commented on samples/extensions/shader_debugprintf/README.adoc:
```diff
@@ -0,0 +1,197 @@
+}
+----
+
+And then gather those so they can be displayed in the graphical user interface. The sample also checks if it's actually a printf message (`pMessageIdName == WARNING-DEBUG-PRINTF`), which is something one could you to distinguish between actual validation messages and printf outputs:
```
typo: one could you -> one could do

asuessenbach commented on samples/extensions/shader_debugprintf/README.adoc:
```diff
@@ -0,0 +1,197 @@
+
+== Printf output buffer size
+
+The default buffer size for debug outputs is limited. If you output a lot of messages, e.g. by doing a printf per fragment shader invocation, this may not suffice and you'll get a validation warning telling you about the buffer being to small. This buffer size can be increased via the VkConfig utility from the LunarG SDK. See https://www.lunarg.com/wp-content/uploads/2021/08/Using-Debug-Printf-02 ...
```
Is there no way to increase that size programmatically?

SaschaWillems replied:
Not that I'm aware of. I didn't find anything in this regards in any of the SDK documents or the LunarG white papers.

asuessenbach commented on samples/extensions/shader_debugprintf/README.adoc:
```diff
@@ -0,0 +1,197 @@
+
+[NOTE]
+====
+This **does not** apply to outputting messages in graphics debuggers.
```
What is "This" here?
That the buffer size is limited? Or that you can increase it via VkConfig?

SaschaWillems replied:
Both. Graphics debuggers don't care for the buffer size and thus changing it with VkConfig has no effect.

asuessenbach commented on samples/extensions/shader_debugprintf/README.adoc:
```diff
@@ -0,0 +1,197 @@
+
+== Memory considerations
+
+Using debug printf will consume a descriptor set, so if you use every last descriptor set it may not work and you may need to increase set count at pool allocation. Thus in the sample we allocate an additional set from the pool. It also consumes device memory based on the no. and/or size fo the debug messages.
```
Maybe: you may need to increase set count -> you may need to increase the set count
Maybe/Typo: on the no. and/or size fo -> on the number and/or size of

asuessenbach commented on samples/extensions/shader_debugprintf/shader_debugprintf.cpp:
```diff
@@ -414,20 +414,18 @@ void ShaderDebugPrintf::create_instance()
+	std::vector<VkValidationFeatureEnableEXT> validation_feature_enables = {VK_VALIDATION_FEATURE_ENABLE_DEBUG_PRINTF_EXT};
+
+	VkValidationFeaturesEXT validation_features{VK_STRUCTURE_TYPE_VALIDATION_FEATURES_EXT};
+	validation_features.enabledValidationFeatureCount = 1;
```
maybe: `1` -> `static_cast<uint32_t>(validation_feature_enables.size())`

asuessenbach commented on samples/extensions/shader_debugprintf/shader_debugprintf.cpp:
```diff
@@ -0,0 +1,500 @@
+void ShaderDebugPrintf::create_instance()
+{
+	// Initialize Volk Vulkan Loader
+	VkResult result = volkInitialize();
```
No need to `volkInitialize` here. It has already been called in `VulkanSample::prepare`.

SaschaWillems replied:
Thanks. The multiple layers of abstraction in the framework always makes it hard to find those things :/

asuessenbach reviewed (commented):
Great description on how to use it.

For some unknown reason, this sample "runs" incredibly slow on my machine (multiple seconds per frame). But when the Vulkan Configurator is running as well, I don't get any output anymore and the speed is as expected. Any idea on that?

SaschaWillems commented:
> Great description on how to use it.
> 
> For some unknown reason, this sample "runs" incredibly slow on my machine (multiple seconds per frame). But when the Vulkan Configurator is running as well, I don't get any output anymore and the speed is as expected. Any idea on that?

That seems to be a bug with the validation layers. I noticed this too, but couldn't reproduce :(

SaschaWillems committed:
Update tutorial

SaschaWillems committed:
Code adjustments based on review

SaschaWillems committed:
Clang-format...

gary-sweet commented:
This no longer crashes for me, but I'm still seeing:

```
[error] [framework/platform/platform.cpp:175] Error Message: Couldn't request feature from device as VK_KHR_get_physical_device_properties2 isn't enabled!
[error] [framework/platform/platform.cpp:176] Failed when running application shader_debugprintf
    71:05:42.547 nxserverlib: Vulkan Display(0x7e5c000ee0) unregistered

```

SaschaWillems commented:
Thank you very much for testing :)

I guess that error is caused by #938

Currently the framework will always enable all features of an extension at the time it's requested. I'll see if I can do a workaround for that.

asuessenbach commented:
I think, @gary-sweet's issue here is, that `VK_KHR_get_physical_device_properties2` is not enabled (is it supported, at all?), even though it has been added in `ShaderDebugPrintf::create_instance`! Should have failed there, then.

gary-sweet commented:
> I think, @gary-sweet's issue here is, that `VK_KHR_get_physical_device_properties2` is not enabled (is it supported, at all?), even though it has been added in `ShaderDebugPrintf::create_instance`! Should have failed there, then.

We do support VK_KHR_get_physical_device_properties2, so it's not obvious what the issue is. I'll do some degugging.

gary-sweet commented:
> I'll do some debugging.

It looks like this sample is creating its own VkInstance which it then constructs the Instance() object from. This appears to bypass a bunch of the extension enabling that happens in the Instance main constructor (all the enable_extension() stuff). The end result is that the Instance object doesn't think the extension is enabled, even though it actually is in the contained VkInstance.

SaschaWillems commented:
Yeah, it's similar to the profiles example. The framework assumes some stuff, even if you create your own instance, which in turn causes problems like these. It also tries to enable all features of an ext structure (see #928). But since I don't want to change anything about the framework in this PR (would be out of it's scope) I'll see if I can workaround this.

SaschaWillems committed:
Merge branch 'main' into shader_printf

SaschaWillems commented:
This is kinda hard to debug, and I feel like this is platform-specific. The only place where this error is thrown is in (hpp_)phyiscal_device.h and at least on Windows I never even enter that code when running my sample. So we probably do something different/specific when running on your platform that causes the assert to trigger. I'll somehow try to reproduce this on Windows.

I guess it's caused by this part of device creation: https://github.com/KhronosGroup/Vulkan-Samples/blob/main/framework/core/device.cpp#L109

It'll implicitly enable some extensions for all samples if they're supported. 

Sometimes I wish our framework wouldn't do so much implicit stuff. Often feels like the exact opposite of Vulkan :/

SaschaWillems committed:
Merge branch 'main' into shader_printf

SaschaWillems commented:
Sorry, this currently fails after merging main. Lots of framework changes and at least something missing for me to continue working on this, see #985,

SaschaWillems committed:
Add new argument to custom instance creation so that extensions enabled in a sample will be marked as enabled in the instance class
This is required to make other parts of the framework properly work with samples that do custom instance creation

SaschaWillems committed:
Add function to check if sample has a valid instance
Fixes #985

SaschaWillems committed:
Adjust to recent framework changes

SaschaWillems committed:
Doc fix

SaschaWillems committed:
Typo

gary-sweet reviewed (changes requested):
I believe this is now missing an entry in samples/CMakeLists.txt. This means the sample isn't currently runnable and doesn't appear in the samples lists in the help.

SaschaWillems commented:
Welp, probably happened while merging with main. Thanks for letting me know, will fix.

SaschaWillems commented:
Can you check again? Afair the list in that cmake file is purely optional. I'm actually not sure why that list of samples in there even exist. CMake should do a folder traversal and adds all samples it can find, no matter if the sample is explicitly added in that list or not.

gary-sweet commented:
> Can you check again? Afair the list in that cmake file is purely optional. I'm actually not sure why that list of samples in there even exist. CMake should do a folder traversal and adds all samples it can find, no matter if the sample is explicitly added in that list or not.

Well that's odd then, because the sample didn't exist for me unless I hacked it into that file.

SaschaWillems commented:
That's interesting. I just reran cmake and deleted the cache beforehand and the sample is somehow picked up. Will have to investigate.

SaschaWillems committed:
Add sample to cmake ordered list

SaschaWillems commented:
Added it to the CMake file anyway. Can't hurt ;)

gary-sweet commented:
> That's interesting. I just reran cmake and deleted the cache beforehand and the sample is somehow picked up. Will have to investigate.

Hmm. So I trashed my build folders completely and rebuilt and then it appeared, without the entry in cmake. It's possible that the cmake generate hadn't run, which is what I was seeing.

gary-sweet reviewed (approved):
The fact that shader-printf 'just works' (provided that you support VK_KHR_shader_non_semantic_info) had kind of passed me by and I had to scratch my head for a little bit when I saw it working with my driver to figure out why!

What a very useful thing, and an excellent sample.

[If I had an extra button to auto-rotate the model it would be even better (since I have no other way to rotate API samples), but I can always add one later on.]
